compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/StreamingCodeDesugar.java

```diff
@@ -840,23 +889,74 @@ private BLangLambdaFunction createAggregatorLambdaWithParams(BLangVariable varAg
-        // (streams:StreamEvent e) => string { .. }
-        BLangLambdaFunction groupingLambda = createLambdaWithVarArg(selectClause.pos, new BLangVariable[]{
-                varGroupByStreamEvent}, symTable.stringType);
-        BLangBlockStmt groupByLambda = groupingLambda.function.body;
-
-        //e.data;
-        BLangIndexBasedAccess mapFieldAccessExpr = createMapAccessExprFromStreamEvent(varGroupByStreamEvent,
-                (BLangExpression) selectClause.getGroupBy().getVariables().get(0));
-        // return check <string>e.data[<fieldName in string>];
-        BLangTypeConversionExpr conversionExpr = generateConversionExpr(mapFieldAccessExpr, symTable.stringType,
-                symResolver);
-        addReturnGroupByFieldStmt(groupByLambda, conversionExpr);
-        return groupingLambda;
+    private ArrayList<BLangExpression> createGroupByLambda(BLangSelectClause selectClause) {
+        ArrayList<BLangExpression> groupByFunctionArray = new ArrayList<>();
+        for (ExpressionNode node : selectClause.getGroupBy().getVariables()) {
+            BLangVariable varGroupByStreamEvent = this.createStreamEventArgVariable(
+                    getVariableName(SELECT_LAMBDA_PARAM_REFERENCE), selectClause.pos, env);
+            // (streams:StreamEvent e) => string { .. }
+            BLangLambdaFunction groupingLambda = createLambdaWithVarArg(selectClause.pos, new BLangVariable[]{
+                    varGroupByStreamEvent}, symTable.stringType);
+            BLangBlockStmt groupByLambda = groupingLambda.function.body;
+
+            //e.data; / getOrderByField(...);
+            BLangTypeConversionExpr conversionExpr = null;
+            if (node.getKind() == NodeKind.FIELD_BASED_ACCESS_EXPR) {
+                BLangIndexBasedAccess mapFieldAccessExpr = createMapAccessExprFromStreamEvent(varGroupByStreamEvent,
+                                                                                              (BLangExpression) node);
+                // return check <string>e.data[<fieldName in string>];
+                conversionExpr = generateConversionExpr(mapFieldAccessExpr, symTable.stringType, symResolver);
+
+            } else if (node.getKind() == NodeKind.INVOCATION) {
+                BLangInvocation invocation = (BLangInvocation) node;
+
+                conversionExpr = generateConversionExpr(refactorInvocationInGroupBy(invocation, varGroupByStreamEvent),
+                                                        symTable.stringType, symResolver);
+            }
+
+            addReturnGroupByFieldStmt(groupByLambda, conversionExpr);
+
+            groupByFunctionArray.add(groupingLambda);
+        }
+        return groupByFunctionArray;
+    }
+
+    private BLangInvocation refactorInvocationInGroupBy(BLangInvocation invocation,
+                                                        BLangVariable varGroupByStreamEvent) {
+        if (invocation.requiredArgs.size() > 0) {
```

Comment:
Without checking if size of the required args is greater than zero at the start, shall we simply return the invocation by checking if size of the args is equal to zero. We can then place the code inside the if condition outside which will improve its clarity. WDYT?

Reply:
That's good. Fixed